src/bundle-manifest.ts

```diff
@@ -150,6 +148,51 @@ export function generateBundles(depsIndex: TransitiveDependenciesMap):
 }
 
 /**
+ * Chains multiple bundle strategy functions together so the output of one
+ * becomes the input of the next and so-on.
+ */
+export function composeStrategies(strategies: BundleStrategy[]) {
+  return strategies.reduce((s1, s2) => (b) => s2(s1(b)));
+}
+
+/**
+ * Generates a strategy function which finds all non-entrypoint bundles which
+ * are dependencies of the given entrypoint and merges them into that
+ * entrypoint's bundle.
+ */
+export function generateEagerMergeStrategy(entrypoint: UrlString):
+    BundleStrategy {
+  return generateMatchMergeStrategy(
+      entrypoint,
+      b => b.entrypoints.has(entrypoint) && !getBundleEntrypoint(b));
+}
+
+/**
+ * Generates a strategy function which finds all bundles matching the predicate
+ * function and merges them into the bundle containing the target file.
+ */
+export function generateMatchMergeStrategy(
+    target: UrlString,
+    predicate: (matchBundle: Bundle, targetBundle: Bundle) =>
+        boolean): BundleStrategy {
+  return (bundles: Bundle[]): Bundle[] => {
+    let targetBundle = bundles.find(bundle => bundle.files.has(target));
+    if (!targetBundle) {
+      throw new Error(`No bundle found containing file ${target}`);
+    }
+    const newBundles = Array.from(bundles);
+    const bundlesToMerge = newBundles.filter(
+        matchBundle => matchBundle === targetBundle ||
+            predicate(matchBundle, targetBundle!));
```

Comment:
For some reason the `if (!targetBundle)` with a throw in it didn't seem to satisfy the type checker and it thought the type was still `Bundle | undefined`, so I `!`'d it.

moot point now, since I don't pass it into that function anymore.